storage/range.go

```diff
@@ -845,27 +892,33 @@ func (r *Range) applyRaftCommand(ctx context.Context, index uint64, originNodeID
 	ms := proto.MVCCStats{}
 
 	// Execute the command; the error will also be set in the reply header.
+	// TODO(tschottdorf,tamird) For #1400, want to refactor executeCmd to not
+	// touch the reply header's error field.
 	intents, err := r.executeCmd(batch, &ms, args, reply)
 
 	if oldIndex := atomic.LoadUint64(&r.appliedIndex); oldIndex >= index {
-		log.Fatalc(ctx, "applied index moved backwards: %d >= %d", oldIndex, index)
+		return nil, newReplicaCorruptionError(
+			util.Errorf("applied index moved backwards: %d >= %d", oldIndex, index))
 	}
 
 	// Advance the applied index atomically within the batch.
 	if e := setAppliedIndex(batch, r.Desc().RaftID, index); e != nil {
-		if reply.Header().GoError() == nil {
-			reply.Header().SetGoError(e)
-		}
-		err = e
+		return nil, newReplicaCorruptionError(
+			util.Errorf("could not update applied index"), e)
+	}
+
+	// If the execution of the command wasn't successful, stop here.
+	if err != nil {
```

Comment:
This can be moved up to immediately after `r.executeCmd`. There's no reason to carry an unexamined error this far. (and then the `setAppliedIndex` block can just use the name `err` instead of dancing around with `err` and `e`).